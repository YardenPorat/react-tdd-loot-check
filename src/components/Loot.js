import React, { Component } from 'react';
import { connect } from 'react-redux';
import { fetchBitcoin } from '../actions/bitcoin';

export class Loot extends Component {
  componentDidMount() {
    this.props.fetchBitcoin();
  }

  calcBitcoinBalance() {
    const { bitcoin } = this.props;

    if (Object.keys(bitcoin).length === 0) return '';

    const rate = parseInt(bitcoin.bpi.USD.rate.replace(',', ''));
    const bitcoinBalance = this.props.balance / rate;
    return bitcoinBalance;
  }

  render() {
    return (
      <div>
        <h3>Bitcoin Balance: {this.calcBitcoinBalance()}$</h3>
      </div>
    );
  }
}

export default connect(state => state, { fetchBitcoin })(Loot);
