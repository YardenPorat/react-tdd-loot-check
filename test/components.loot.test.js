import React from 'react';
import { shallow, mount } from 'enzyme';
import { Loot } from '../src/components/Loot';
import { expect } from 'chai';
import sinon from 'sinon';

describe('Loot Component', () => {
  let props = { balance: 10, bitcoin: {}, fetchBitcoin: () => {} };
  let loot = shallow(<Loot {...props} />);

  it('renders properly', () => {
    expect(loot).to.matchSnapshot();
  });

  describe('when mounted', () => {
    let fetchBitcoinStub;
    var sandbox;

    beforeEach(() => {
      sandbox = sinon.createSandbox();
    });

    afterEach(() => {
      sandbox.restore();
    });

    it('should activate componentDidMount', () => {
      sandbox.spy(Loot.prototype, 'componentDidMount');
      loot = mount(<Loot {...props} />);
      expect(Loot.prototype.componentDidMount).to.have.property('callCount', 1);
    });

    it('dispatches the `fetchBitcoin()` method from props', () => {
      fetchBitcoinStub = sandbox.stub(props, 'fetchBitcoin');
      loot = mount(<Loot {...props} />);
      // eslint-disable-next-line no-unused-expressions
      expect(fetchBitcoinStub).to.have.been.calledOnce;
      expect(fetchBitcoinStub).to.have.been.calledWith(); //nothing
    });
  });

  describe('when there are valid bitcoin props', () => {
    beforeEach(() => {
      props = {
        balance: 10,
        bitcoin: { bpi: { USD: { rate: '1,000' } } },
        fetchBitcoin: () => {},
      };
      loot = shallow(<Loot {...props} />);
    });

    it('displays the correct bitcoin value', () => {
      expect(loot.find('h3').text()).to.eql('Bitcoin Balance: 0.01$');
    });
  });
});
