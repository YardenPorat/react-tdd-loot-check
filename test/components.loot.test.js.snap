// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loot Component renders properly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Loot
    balance={10}
    bitcoin={Object {}}
    fetchBitcoin={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <h3>
        Bitcoin Balance: 
        
        $
      </h3>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          "Bitcoin Balance: ",
          "",
          "$",
        ],
      },
      "ref": null,
      "rendered": Array [
        "Bitcoin Balance: ",
        "",
        "$",
      ],
      "type": "h3",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <h3>
          Bitcoin Balance: 
          
          $
        </h3>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Bitcoin Balance: ",
            "",
            "$",
          ],
        },
        "ref": null,
        "rendered": Array [
          "Bitcoin Balance: ",
          "",
          "$",
        ],
        "type": "h3",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSeventeenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;
